src/encoding.ts

```diff
@@ -0,0 +1,379 @@
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function normalizeEncodingArgs(data: any, alternateCodec: any): [ string, ByteBuffer ]{
+	data = String(data);
+
+	return [ data, <ByteBuffer> {
+		toString: function (codec?: any): string {
+			return (codec || alternateCodec).decode(this);
+		},
+		toJSON: function (): string {
+			return JSON.stringify(Array.prototype.slice.call(this));
+		},
+		length: data.length
+	}];
+}
+
+function validateDecodingArgs(data: any) {
+	if (data == null) {
+		return false;
+	}
+
+	if (data.length == null) {
+		throw new TypeError('Argument must have a length property');
+	}
+
+	return true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [0, Infinity], checkSurrogate?: boolean): string {
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
+		throw Error('Surrogate is not a scalar value');
+	}
+
+	let encoded = '';
+
+	if (codePoint > 0xFFFF) {
+		codePoint = codePoint - 0x010000;
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+export interface ByteBuffer {
+	[index: number]: number;
+	toString?: (codec?: any) => string;
+	toJSON?: () => string;
+	length: number;
+}
+
+/**
+ * Ascii codec that provides facilities for encoding a string (or string coerced object)
+ * into an Ascii encoded byte buffer as well as decoding an Ascii encoded byte buffer
+ * into a string.
+ */
+export class Ascii {
```

Comment:
I think the benefits of using 2-function modules with an interface outweigh the ability to subclass. Having an interface will allow these to be easily passed into functions, which will be a key use of codecs. If someone wants to extend one of these codecs, they can just wrap it and call the wrapped object's methods.